src/main/java/com/marklogic/spark/reader/optic/OpticReadContext.java

```diff
@@ -130,20 +172,55 @@ private PlanBuilder.Plan buildPlanForBucket(RowManager rowManager, PlanAnalysis.
             .bindParam("ML_LOWER_BOUND", bucket.lowerBound)
             .bindParam("ML_UPPER_BOUND", bucket.upperBound);
 
+        // Add user-defined parameters from options
+        for(Map.Entry<String, String> entry : this.userDefinedParams.entrySet()) {
+            if (logger.isDebugEnabled()) {
+                logger.debug("Adding user param: {}", entry);
+            }
+            plan = plan.bindParam(entry.getKey(), entry.getValue());
+        }
+
         if (opticFilters != null) {
             for (OpticFilter opticFilter : opticFilters) {
                 plan = opticFilter.bindFilterValue(plan);
             }
         }
 
+        for (Map.Entry<String, String> entry : columnParamValues.entrySet()) {
+            if (entry.getValue() == null) {
+                entry.setValue(columnParamDefaults.get(entry.getKey()));
+                if (logger.isDebugEnabled()) {
+                    logger.debug("Setting default param column: {}", entry);
```

Comment:
I think `trace` would be more appropriate for this logging. It'll get executed for every bucket and it will log the same thing every time, which is very noisy. Also, logging user-provided values can be risky - e.g. this could be a sensitive value which should not be logged. I lean towards not having this logging for now for that reason, and after we use this feature for a bit we can determine if it's really needed - perhaps we only log the param name.

Reply:
Most of the logging I added was because I was debugging something in the Java code, not because I needed it as a user, so a lot of it is probably excessive.